cdc/metadata/metadata.go

```diff
@@ -0,0 +1,169 @@
+
+// ---------- About changefeed worker captures scheduling ---------- //
+// 1. ControllerObservation.AttachWorkers attaches some captures to a changefeed;
+// 2. Owner calls OwnerObservation.RefreshWorkers to know how to launch or stop processors;
+// 3. Capture receives processor launch/stop messages with P2P mechanism;
+// 4. Capture doesn't care anything about changefeed worker captures scheduling.
+// 5. How to rolling-update a changefeed with only one worker capture:
+//    * controller needs only to attach more captures to the changefeed;
+//    * it's owner's responsibility to evict tables between captures;
+// ---------------------------------------------------------------- //
+
+// ---------------- About keep-alive and heartbeat ---------------- //
+// 1. Capture updates heartbeats to metadata by calling CaptureObservation.Heartbeat,
+//    with a given timeout, for example, 1s;
+// 2. On a capture, controller and owners share one same Context, which is associated
+//    with deadline 10s. CaptureObservation.Heartbeat will refresh the deadline.
+// 3. Controller is binded with a lease (10+1+1)s, for deadline, heartbeat time-elapsed
+//    and network clock skew.
+// 4. Controller needs to consider re-scheduling owners from a capture, if the capture
+//    has been partitioned with metadata storage more than lease+5s;
+// 5. Controller needs to consider re-scheduling changefeed worker captures, if a
+//    changefeed worker capture has been missed more than lease+5s, and changefeed lag
+//    is larger than lease+5s.
+// ---------------------------------------------------------------- //
+
+// CaptureObservation is for observing and updating metadata on a CAPTURE instance.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type CaptureObservation interface {
+	// SelfCaptureInfo tells the caller who am I.
+	SelfCaptureInfo() *CaptureInfo
+
+	// Heartbeat tells the metadata storage I'm still alive.
+	Heartbeat(context.Context) error
+
+	// TakeControl blocks until becomes controller or gets canceled.
+	TakeControl() ControllerObservation
+
+	// Advance is like OwnerObservation.Advance but handles many changefeeds.
+	//
+	// It should returns an InvalidOwner error if any changefeed ownership is revoked.
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest alive changefeed owner list from metadata.
+	RefreshAliveOwners() <-chan []*ChangefeedInfo
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
+	// CreateChangefeed creates a changefeed.
+	CreateChangefeed(cf *ChangefeedInfo, up *UpstreamInfo) error
+
+	// UpdateChangefeed updates changefeed metadata, must be called on a stopped one.
+	UpdateChangefeed(cf *ChangefeedInfo) error
+
+	// Stop a changefeed.
+	SetChangefeedStopped(cf *ChangefeedInfo) error
```

Comment:
Duplicate with UpdateChangefeed? In addition, we should restrict the cahngefeed state only at the api level.